Metaformer baselines for vision

fffffgggg54 opened the pull request:
PR adapting metaformers from https://github.com/sail-sg/metaformer to the main timm codebase.
Implements identityformer, randformer, poolformerv2, convformer, and caformer from https://arxiv.org/abs/2210.13452 and integrates poolformer (v1) from https://arxiv.org/abs/2111.11418.

Known issues:
- Randformer fails tests (not all parameters require gradients)
- Weights saved from old poolformer impl don't work
- Current checkpoint mapping is messy, should get changed to use `group_matcher`
- Excessive permute statements. Having everything channels-last was a workaround for torchscript. Should now be unnecesary following some other changes.
- Randformer only works on fixed input size (other models allow for arbitrary sizes)

I'd like to hear any additional feedback before proceeding to address these issues. My validation results line up with the paper's.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/timm/pr_1647). All of your documentation changes will be reflected on that endpoint.

fffffgggg54 commented:
> I've converted models with unused params before, have you identified what the params are and why the authors might have not realized they weren't used? they should get removed for a cleaner model
> 
RandFormer has a matrix that maps input tokens to output tokens at random. It doesn't require grad, since it isn't meant to be updated during training but should still be included in saved checkpoints. The module is at https://github.com/fffffgggg54/pytorch-image-models/blob/metaformer_baselines_for_vision/timm/models/metaformers.py#L154.

For the checkpoint mapping, the current scheme works for pretrained weights that are loaded during the `timm.create_model()` call. My concern is that, in order to merge poolformer (v1) into this new impl, old checkpoints from the current poolformer impl that are saved by the user need to be mapped to the new model. Is there a way to support this or does the old impl need to stay for legacy support?
Ex:
```python
pf_original = timm.create_model('poolformer_s12')
torch.save(pf_original.state_dict(), 'original.pth')

# Doesn't work
pf_new = timm.create_model('poolformerv1_s12')
pf_new.load_state_dict(torch.load('original.pth'))
```

rwightman commented:
@fffffgggg54 there's a lot going in this impl, some time to mentally parse

Looking at it, it looks 'slow' with so many permutes, the custom layer norm is **very** slow w/o using torch.compile (PyTorch 2.0) ass well.  I wonder if doing everything w/ convolutions would make more sense, linear tends to be faster as you scale up the width but these models aren't super large. Hmm.

For the random mixing, that should be a buffer not a parameter since it looks like it's supposed to be fixed.

fffffgggg54 commented:
Changed to channels first, randformer uses buffer and allows for arbitrary input sizes, still use einsum and run feature extraction test for randformer.

fffffgggg54 commented:
draft for now, convformer and caformer completely fail validation (~55% and ~0.02%), other 4 variants work.

fffffgggg54 commented:
regression on a189593b02ea811058eafd1545a6d06f02f9b42b

fffffgggg54 commented:
The regression after the channels-first change had to do with the norm layers used. This has been fixed. The BCHW randformer implementation was also updated to match val results from paper, problem had to do with reshaping. I added an interpolationg op in the random mixing module. The model has the same val results at when run at train res, but getting actual performance out of it beyond slight variations and reductions, in eval size still wip. Change allows for varied input sizes should someone choose to do so with the model. Some other parts of the implementation may still need a bit of cleanup.

@yuweihao do you have any input? I know you mentioned wanting to have the new poolformer impl replace the old one, with the model renamed to `poolformer_*`. This would break backwards compatibility with weights saved from the old model definition, but since there is a new version of timm planned with backwards-compatability-breaking changes, I'm not sure if this is an issue.

<details><summary>validation results of channels-first are within error of ea054117ee65bd0ed1da16cbd7ffb7e15c3cf521 and the paper.</summary>

```
new:
model: identityformer_s12.sail_in1k, top-1: 74.58%, spent 721 seconds
model: randformer_s12.sail_in1k, top-1: 76.63799999999999%, spent 721 seconds
model: poolformerv1_s12.sail_in1k, top-1: 77.14999999999999%, spent 721 seconds
model: poolformerv2_s12.sail_in1k, top-1: 77.996%, spent 721 seconds
model: convformer_s18.sail_in1k, top-1: 82.98%, spent 721 seconds
model: caformer_s18.sail_in1k, top-1: 83.66799999999999%, spent 721 seconds

old:
model: identityformer_s12.sail_in1k, top-1: 74.58%, spent 699 seconds
model: randformer_s12.sail_in1k, top-1: 76.63%, spent 699 seconds
model: poolformerv1_s12.sail_in1k, top-1: 77.146%, spent 699 seconds
model: poolformerv2_s12.sail_in1k, top-1: 78.006%, spent 699 seconds
model: convformer_s18.sail_in1k, top-1: 82.98400000000001%, spent 699 seconds
model: caformer_s18.sail_in1k, top-1: 83.662%, spent 699 seconds
```
</details>

<details><summary>nn.Linear ->1x1 conv results in noticeable (17-20%) performance gain</summary>

```
identityformer_s12, tesla m40, bs=64, 1x1 conv: ~155.5 ips
identityformer_s12, tesla m40, bs=64, fc: - ~129.1 ips

identityformer_m48, tesla m40, bs=16, 1x1 conv: - ~27.6 ips
identityformer_m48, tesla m40, bs=16, fc: - ~23.45 ips
```
</details>

<details><summary>224 train -> new test resolution, randformer interpolation performs poorly, other models as expected</summary>

```
eval @ 128:
model: identityformer_s12.sail_in1k, top-1: 63.94800000000001%, spent 455 seconds
model: randformer_s12.sail_in1k, top-1: 64.862%, spent 455 seconds
model: poolformerv1_s12.sail_in1k, top-1: 65.684%, spent 455 seconds
model: poolformerv2_s12.sail_in1k, top-1: 67.22%, spent 455 seconds
model: convformer_s18.sail_in1k, top-1: 75.8%, spent 455 seconds
model: caformer_s18.sail_in1k, top-1: 75.68%, spent 455 seconds

eval @ 160:
model: identityformer_s12.sail_in1k, top-1: 69.652%, spent 528 seconds
model: randformer_s12.sail_in1k, top-1: 71.544%, spent 528 seconds
model: poolformerv1_s12.sail_in1k, top-1: 72.25200000000001%, spent 528 seconds
model: poolformerv2_s12.sail_in1k, top-1: 73.262%, spent 528 seconds
model: convformer_s18.sail_in1k, top-1: 79.852%, spent 528 seconds
model: caformer_s18.sail_in1k, top-1: 80.31%, spent 528 seconds

eval @ 288:
model: identityformer_s12.sail_in1k, top-1: 75.47399999999999%, spent 1022 seconds
model: randformer_s12.sail_in1k, top-1: 75.31400000000001%, spent 1022 seconds
model: poolformerv1_s12.sail_in1k, top-1: 78.082%, spent 1022 seconds
model: poolformerv2_s12.sail_in1k, top-1: 79.30199999999999%, spent 1022 seconds
model: convformer_s18.sail_in1k, top-1: 83.572%, spent 1022 seconds
model: caformer_s18.sail_in1k, top-1: 83.72%, spent 1022 seconds

eval @ 384:
model: identityformer_s12.sail_in1k, top-1: 74.76%, spent 1815 seconds
model: randformer_s12.sail_in1k, top-1: 1.8599999999999999%, spent 1815 seconds
model: poolformerv1_s12.sail_in1k, top-1: 77.63799999999999%, spent 1815 seconds
model: poolformerv2_s12.sail_in1k, top-1: 78.956%, spent 1815 seconds
model: convformer_s18.sail_in1k, top-1: 82.178%, spent 1815 seconds
model: caformer_s18.sail_in1k, top-1: 81.582%, spent 1815 seconds
```
</details>

yuweihao commented:
Hi @fffffgggg54 , many thanks for your contribution.

Could I give some small suggestions? It is better that your contributed `metaformers.py` does not contain old poolformer. That is, it is better to remove poolformerv1 in `metaformers.py`. This way may make the code have better compatibility. 

Besides, it is better that `metaformers.py` is renamed to `metaformer_baselines.py`. The first reason is that metaformer is an abstract concept while metaformer baselines are specific models. The second reason is that it can remind users that old poolformer is in `poolformer.py` and poolformerv2 is in `metaformer_baselines.py`.

Thank you very much again for your contribution.

fffffgggg54 commented:
IMO for the old version of poolformer, it makes more sense to merge the architecture into the metaformer definition from an organizational standpoint. Since all the code is repeated and instantiating old poolformer requires only a few kwarg changes, having a separate definition feels redundant, especially since throughput-wise, they perform identically on my setup.

For this same reason, I believe that the file should remain as `metaformers.py`, as the structure is general and can be extended to other configurations not outlined within the baselines. Any model fitting within the metaformer concept can probably be implemented by instantiating a modified version of the `MetaFormer` module, as long as they don't use cls tokens. In practice, things like CoatNet/MaxViT/MOAT, while closely related, would require making the implementation more complicated (stem changes, positional embedding, more flexible blocks, etc).

Ultimately, @rwightman has the final say, delete `poolformer.py`, rename `poolformerv1_*` to `poolformer_*`, possibly write something to remap old weights to new when trying to load weights, or remove `poolformerv1_*` and/or rename the file to `metaformer_baselines.py`.

fffffgggg54 commented:
The current implementation of the no-bias norm layers doesn't work with fast norm. Not sure if there is a better way to go about it though. Compilation and functional SDPA seem to have sped up caformer significantly.

rwightman commented:
@fffffgggg54 thanks for the update, re the no-bias fast norm, unless the kernels can't handle it, pretty sure if this line in timm https://github.com/huggingface/pytorch-image-models/blob/main/timm/layers/fast_norm.py#L81 is changed to check bias it would work (assuming bias is forced to None like your override)

fffffgggg54 commented:
changed fast norm, both ln2d and gn1 models work

fffffgggg54 commented:
@rwightman yeah, go ahead

rwightman commented:
Merged via #1793 , there was some fixes (some models using wrong norm layer, and cleanup to do, I removed the ident/rand models as they're more of a 'paper' model than useful weights for others.
